---
title: "TCGA BRCA Data Exploration"
output: html_notebook
---

## Setup

```{r}
library(tidyverse)
library(synapser)
library(GenomicDataCommons)
library(stringr)
library(janitor)
library(ggthemes)
library(patchwork)
library(ggalluvial)
library(forcats)
library(kableExtra)
```

## Data collection

### Clinical data

```{r}
# load TCGA PanCanAtlas manifest for GDC
manifest_df <- read_tsv("../data/PanCan-General_Open_GDC-Manifest_0.txt")
```

```{r, warning=FALSE, include=FALSE}
tcga_cdr_clinical_file <- manifest_df %>% 
  dplyr::filter(filename == "TCGA-CDR-SupplementalTableS1.xlsx") %>% 
  pluck("id") %>% 
  gdcdata(destination_dir = "../data/")
```

```{r, warning=FALSE}
brca_cdr_clinical_df <- readxl::read_excel(tcga_cdr_clinical_file) %>% 
    dplyr::filter(type == "BRCA")

sparse_vars <- brca_cdr_clinical_df %>%
  summarise_all(.funs = funs(sum(str_detect(., "^\\[Not A.*") | is.na(.)))) %>% 
  gather(variable, n_uninformative) %>% 
  dplyr::filter(n_uninformative > (nrow(brca_cdr_clinical_df) * 0.25)) %>% 
  pluck("variable")

brca_cdr_clinical_df <- dplyr::select(brca_cdr_clinical_df, -one_of(sparse_vars))
```

```{r, warning=FALSE}
tcga_clinical_file <- manifest_df %>% 
  dplyr::filter(filename == "clinical_PANCAN_patient_with_followup.tsv") %>% 
  pluck("id") %>% 
  gdcdata(destination_dir = "../data/")
```

```{r, warning=FALSE}
brca_clinical_df <- read_tsv(tcga_clinical_file) %>% 
  dplyr::filter(acronym == "BRCA")

sparse_vars <- brca_clinical_df %>%
  summarise_all(.funs = funs(sum(str_detect(., "^\\[Not A.*") | is.na(.)))) %>% 
  gather(variable, n_uninformative) %>%  
  dplyr::filter(n_uninformative > (nrow(brca_clinical_df) * 0.25)) %>% 
  pluck("variable")

brca_clinical_df <- dplyr::select(brca_clinical_df, -one_of(sparse_vars))
```


### Expression data

```{r, warning=FALSE, include=FALSE}
tcga_expr_file <- manifest_df %>% 
  dplyr::filter(
    filename == "EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv"
  ) %>% 
  pluck("id") %>% 
  gdcdata(destination_dir = "../data/")
```


#### Retrieve/load mRNA data

Batch effects normalized mRNA data: syn4976363
syn4976369 ("EB++AdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv")

```{r include=FALSE}
brca_expr_norm_file <- "../data/tcga_brca_expr_norm_df.RData"
if (!fs::file_exists(brca_expr_norm_file)) {
  # download and load normalized, batch-corrected expression data
  tcga_expr_norm_file <- synGet("syn4976369", downloadLocation = "../data/",
                                ifcollision = "overwrite.local")
   tcga_expr_norm_df <- read_tsv(tcga_expr_file$path, progress = FALSE)
  
  # load IRWG feature matrix for barcode mapping
  irwg_fmx_file <- synGet("syn11187757",  downloadLocation = "../data/",
                          ifcollision = "overwrite.local")
  load(irwg_fmx_file$path)
  irwg_fmx_df <- df %>% 
    mutate_if(is.factor, as.character)
  rm(df)
  
  brca_barcode_map <- irwg_fmx_df %>% 
    dplyr::filter(Study == "BRCA") %>% 
    dplyr::select(
      ParticipantBarcode,
      AliquotBarcode = Representative_Expression_Matrix_AliquotBarcode
    ) %>% 
    dplyr::filter(AliquotBarcode %in% names(tcga_expr_norm_df)) %>% 
    deframe()
  
  # subset expression data with only BRCA samples
  brca_expr_norm_df <- tcga_expr_norm_df[, c("gene_id", brca_barcode_map)] %>% 
    dplyr::rename(!!!brca_barcode_map) %>% 
    tidyr::separate(gene_id, c("symbol", "entrez"), sep = "\\|")
  
  # update/augment gene identifiers
  gene_df <- brca_expr_norm_df %>% 
    dplyr::select(entrez) %>% 
    mutate(entrez = as.integer(entrez)) %>% 
    left_join(grch38, by = "entrez") %>% 
    dplyr::select(gene_id = ensgene, entrez, symbol, biotype, description)
  
  # remove genes with no Ensembl ID
  brca_expr_norm_df <- brca_expr_norm_df  %>% 
    dplyr::select(-symbol) %>% 
    mutate(entrez = as.integer(entrez)) %>% 
    left_join(dplyr::select(gene_df, gene_id, entrez), by = "entrez") %>% 
    dplyr::select(gene_id, -entrez, everything()) %>% 
    dplyr::filter(!is.na(gene_id))
  gene_df <- gene_df %>% 
    dplyr::filter(gene_id %in% brca_expr_norm_df$gene_id)
  
  save(brca_expr_norm_df, file = brca_expr_norm_file)
  save(gene_df, file = "../data/tcga_brca_expr_gene_info.RData")
} else {
  load(brca_expr_norm_file)
  load("../data/tcga_brca_expr_gene_info.RData")
}

```

## Data formatting

## Exploratory data analysis

```{r}
brca_clinical_df %>% 
  filter_at(.vars = vars(matches("pathologic_[M|N]")), 
            .vars_predicate = all_vars(!str_detect(., "X"))) %>%
  mutate(
    path_tmn_n = if_else(str_detect(pathologic_N, "N0"), "N0", "non-N0"),
    path_tmn_m = if_else(str_detect(pathologic_M, "M0"), "M0", "non-M0")
  ) %>% 
  dplyr::select(path_tmn_n, path_tmn_m) %>%
  group_by(path_tmn_n, path_tmn_m) %>%
  tally() %>%
  ungroup() %>%
  rownames_to_column("alluvium") %>% 
  gather(tmn_category, tmn_stage, path_tmn_n, path_tmn_m) %>%
  mutate(
    tmn_group = if_else(str_detect(tmn_stage, "non"), "non-0", "0"),
    tmn_category = fct_inorder(tmn_category)
  ) %>%
  ggplot(aes(x = tmn_category, stratum = tmn_stage, alluvium = alluvium, 
             weight = n, fill = tmn_group, label = tmn_stage)) +
  geom_flow(colour = "black", width = 1/5) +
  geom_stratum(alpha = 0.7, width = 1/5) +
  geom_label(aes(colour = tmn_group), stat = "stratum", size = 3) +
  scale_x_discrete(expand = c(0, 0)) +
  scale_fill_colorblind() + 
  scale_colour_manual(values = c("white", "black")) +
  guides(fill = FALSE, colour = FALSE) + 
  theme_minimal()
```

```{r}

```

