---
title: "Understanding data in reduced dimensions"
author: "Diana Murray"
date: "July 12, 2018"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)

set_basepath <- function(type = c("data", "R")) {
  if (stringr::str_length(Sys.which("rstudio-server"))) {
    file.path("/home/shared", type)
  } else {
    here::here(type)
  }
}

data_dir <- set_basepath("data")
scripts_dir <- set_basepath("R")
source(file.path(scripts_dir, "submission_helpers.R"))
```

# About this activity

We will use the R package **clusterProfiler** to perform enrichment analysis of GO annotations for sets of genes we think may be related to motility (more about this below). We know from this course, our research, articles, and other courses we may have taken, that we most likely do not need to write our own code to perform most bioinformatics analyses. But how do we know where to find trustworthy code?

The clusterProfiler package is maintained through [**Bioconductor**](https://bioconductor.org/). Bioconductor provides tools for the analysis and comprehension of high-throughput genomic data. Bioconductor uses the R statistical programming language, and is open source and open development. A large community of professionals supports its development, maintenance, and ease of use. There are a number of ways to install Bioconductor packages into your local RStudio environment (https://bioconductor.org/install/).

Here is a paper that describes the project and provides a tutorial: https://www.ncbi.nlm.nih.gov/pubmed/25633503.

At the top of the [Bioconductor page for clusterProfiler](http://bioconductor.org/packages/release/bioc/html/clusterProfiler.html), there are some metrics that describe how widely used and trustworthy the package is — e.g., its download profile (in top 5%), the support activity for the last six months, and a multi-platform check report. This package itself is also supported by a publication: https://www.ncbi.nlm.nih.gov/pubmed/22455463. 

If you scroll down the clusterProfiler Bioconductor page, there are links to a manual and vignettes (working examples of the package with data provided. "Vignettes" are an outstanding feature of Bioconductor packages.  Overall, clusterProfiler is a solid package.

When we install a package on our local machines, we have to call them with the **`library()`** function.

```{r, message=FALSE, warning=FALSE}
library(clusterProfiler)
library(org.Hs.eg.db)
library(DOSE)
```

---

# Loading & formatting data

We revist the work we did with Andrew in **Module 1** and load and manipulate the PS-ON cell line expression and motility data.

We implement a modification by taking care to capture the Ensembl IDs for the genes.  We avoid most of the 1-to-many conversion issues that can arise with other nomenclatures. For example, human KRAS has the Ensembl id: "ENSG00000133703"; the hugo gene ID: KRAS; the UniProt ID: P01116; and the Entez ID: 3845. 

Run the following chunk to create the expression matrix for 18K+ genes across the ~60 cell lines.

```{r, warning=FALSE}
load(file.path(data_dir, "pson_expr_tpm_df.RData"))
load(file.path(data_dir, "pson_expr_gene_info.RData"))
load(file.path(data_dir, "pson_motility_tidy_df.RData"))

pson_expr_tpm_df2 <- merge(
  x = gene_df, y = pson_expr_tpm_df,
  by.x = "gene_id", by.y = "gene_id"
)

pson_tpm_mat <- as.matrix(pson_expr_tpm_df2[, -c(1:5)])

pson_tpm_mat_ensembl <- pson_tpm_mat
rownames(pson_tpm_mat_ensembl) <- pson_expr_tpm_df2$gene_id

pson_logtpm_mat_ensembl <- log2(1 + pson_tpm_mat_ensembl)

pson_logtpm_mat_ensembl[1:5, 1:5]
```

As we did in Module 1, we will use the data for the hyaluronic acid collagen substrate and select only the samples representing breast cancer cell lines.

```{r, warning=FALSE}
cell_speeds_df <- subset(pson_motil_tidy_df, summary_metric == "speed_um_hr")

hyal_coll_df <- subset(
  cell_speeds_df,
  experimentalCondition == "HyaluronicAcid Collagen"
)

samples_w_expr <- match(hyal_coll_df$sample, colnames(pson_logtpm_mat_ensembl))

hyal_coll_logtpm_mat_ensembl <- pson_logtpm_mat_ensembl[, samples_w_expr]

dim(hyal_coll_logtpm_mat_ensembl)

which(hyal_coll_df$diagnosis == "Breast Cancer")
```

---

# Associating genes with differences in motility

In this chunk, we create a data frame that contains the ranked difference in expression between the two breast cancer cell lines, **T-47D** and **MDA-MB-231**.  Very different average speeds are observed on the hyaluronic acid collagen substrate: 16.5 u/hr for T-47D and 36.2 u/hr for MDA-MB-231.

```{r, warning=FALSE}
breast_logtpm_df <- data.frame(
  rownames(hyal_coll_logtpm_mat_ensembl),
  hyal_coll_logtpm_mat_ensembl[, c(7, 9)]
)

names(breast_logtpm_df) <- c("ensembl", "slower", "faster")

breast_logtpm_df$delta <- breast_logtpm_df[, "faster"] -
  breast_logtpm_df[, "slower"]
breast_logtpm_df <- breast_logtpm_df[order(breast_logtpm_df$delta, decreasing = TRUE), ]

rownames(breast_logtpm_df) <- NULL

head(breast_logtpm_df, 10)
```

The genes at the top of this list show much greater expression in the *faster* cell line; if we had ordered the list by `decreasing = FALSE`, then we'd have negative delta values at the top and be looking at genes with greater expression in the *slower* cell line.

We won't use the motility measures below, so let's just pull out the gene IDs and `delta` values.

```{r, warning=FALSE}
breast_logtpm_df2 <- breast_logtpm_df[, -c(2, 3)]

head(breast_logtpm_df2, 10)
tail(breast_logtpm_df2, 10)
```

In this chunk, we set up the gene list we need for performing the GO enrichment analysis.

`geneList_pson` is simply a ranked list of "delta" values (the difference in expression for each gene in the "fast" versus "slow" condition) with the Ensembl gene IDs as names. 

```{r, warning=FALSE}
geneList_pson <- breast_logtpm_df2[, 2]
names(geneList_pson) <- breast_logtpm_df2[, 1]

head(geneList_pson)
```

With Andrew in Module 1, we inspected the genes at the top and bottom of this ranked list. Andrew noted that some of the genes most differentially expressed were lotto winners. :)

To focus on the genes with the strongest apparent relationship to motility, we'll select only those at the top or bottom of the list. Choosing a cutoff in these situations can often be a bit arbitrary... We can use the `hist()` function to check the distribution of `delta` values and try to pick a threshold that includes a reasonable fraction of the "extreme" values.

```{r}
hist(geneList_pson)
abline(v = c(-4, 4), lty = 2)
```

Again, somewhat arbitrary, but we'll go ahead and use 4 as our cutoff for `delta` in both directions.

We create two sets of genes: 1) Those with "delta" less than or equal to -4 (bottom of the list), and 2) those with "delta" greater than or equal to +4 (top of the list).


```{r, warning=FALSE}
gene_pson_neg <- names(geneList_pson[(geneList_pson <= -4)])
gene_pson_pos <- names(geneList_pson[(geneList_pson >= 4)])
```

---

# Gene Ontology (GO) enrichment

The **clusterProfiler** package provides functions for performing a wide range of enrichment analyses.  We will focus on **`enrichGO`**, which annotates the genes in our list with GO terms and then looks for GO annotations that are significantly over- or under-represented.

Look are carefully at the arguments to the function. Note that the `pAdjustMethod` is set to `"BH"`. The Benjamini-Hochberg (BH) procedure is a powerful statistical tool that decreases the false discovery rate in analysis. Adjusting the rate helps to control for the fact that sometimes small p-values (less than 5%) happen by chance, which could lead us to incorrectly rejecting the true null hypotheses. In other words, the BH Procedure helps us avoid Type I errors (false positives).

The ["Statistics How To" site](http://www.statisticshowto.com/benjamini-hochberg-procedure/) offers some nice descriptions of statistical concepts.

What are we using as our background gene set or "universe"?  Why this list and not another?

Check out the documentation for the `enrichGO()` function to learn a bit more about how it works.

```{r}
?enrichGO
```

Now we'll actually run the function and check out the results.

```{r, warning=FALSE}
ego_pson_pos <- enrichGO(
  gene = gene_pson_pos,
  universe = names(geneList_pson),
  OrgDb = org.Hs.eg.db,
  keytype = "ENSEMBL",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_pos)
```

Are these annotations consistent with our "hypothesis" that genes with the greatest differential expression between the "fast" and "slow" breast cancer cell lines may be related to motility?

A more helpful, comprehensive way to inspect your results is with the visualization tools **clusterProfile** provides.  Run the following chunk.  What do the plots tell you?  Do you prefer one?

```{r}
barplot(ego_pson_pos)
dotplot(ego_pson_pos)
```


Let's look at the enrichment of annotations for genes at the bottom of the rank list — i.e. those genes that are under-expressed in "fast" versus "slow" breast cancer cell lines.

```{r, warning=FALSE}
ego_pson_neg <- enrichGO(
  gene = gene_pson_neg,
  universe = names(geneList_pson),
  OrgDb = org.Hs.eg.db,
  keytype = "ENSEMBL",
  ont = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_neg)
```

Recall that GO has three main aspects:

1. biological process (BP),
2. molecular function (MF),
3. cellular component (CC). 

We already looked at biological process (BP) annotations in the previous two examples.  Use the `enrichGO()` function, provided again in the next chunk, to examine the enrichment of "MF" and "CC" annotations.  

```{r, warning=FALSE}
ego_pson_pos <- enrichGO(
  gene = gene_pson_pos,
  universe = names(geneList_pson),
  OrgDb = org.Hs.eg.db,
  keytype = "ENSEMBL",

  # to change the GO aspect, change the ont argument.  It can be "BP", "MF", or "CC"

  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_pos)
```

Use barplot or dotplot to visualize the enrichments.

```{r}
# visualize your results with either the barplot() or dotplot() function

```

---

# mini-DREAM Challenge

**QUESTION:** How does enrichment analysis look when we use the ranked absolute values of delta? Consider BP, MF, or CC.

First, set up the data to use as input for the `enrichGO` function.

```{r, warnings=FALSE}
breast_logtpm_df3 <- breast_logtpm_df2

# Replace 'aaa' with a function that will return the absolute values of delta.
# HINT: 'a' is the first letter of the function name...
breast_logtpm_df3[, 2] <- aaa(breast_logtpm_df3[, 2])

# Replace ?? with T or F.  How should the delta's be ranked? Renen
breast_logtpm_df3 <- breast_logtpm_df3[order(breast_logtpm_df3$delta, decreasing = ??), ]
head(breast_logtpm_df3)

geneList_abs <- breast_logtpm_df3[, 2]
names(geneList_abs) <- breast_logtpm_df3[, 1]

# Replace nnnnn with the function that will provide the information you need.
# HINT: 'n' is the first letter of the function name...
gene_pson_abs <- nnnnn(geneList_abs[(geneList_abs >= 4)])
```

Check the length of the resulting gene list, `gene_pson_abs` (either using a command you've seen previously or by looking in the environment panel).

```{r}


# fill in the number of genes in the list
my_gene_count <- 0
```

Now, pick a GO aspect — or "subontology" — and perform the enrichment analysis. Inspect the results and answer a few questions below.

```{r}
# set the GO aspect you want to use with `my_subontology`;
# fill in "BP", "MF", or "CC"
my_go_subontology <- "CC"

# Use one of the enrichGO implementations above to perform the analysis
ego_pson_abs <- enrichGO(
  gene = gene_pson_abs,
  universe = names(geneList_pson),
  OrgDb = org.Hs.eg.db,
  keytype = "ENSEMBL",
  ont = my_go_subontology,
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_abs)
```

What did you find to be the most significantly enriched GO term in your list of genes, based on adjusted p-value?

```{r}
# fill in the GO term ID (e.g., "GO:0012345")
my_top_go_id <- "GO:"

# fill in the description of the GO term 
# (you can copy/paste this from the table)
my_go_description <- ""
```


Use one of the visualization tools from above to interpret your results.

```{r}
# replace ___ in the function name to create your plot
___plot(ego_pson_abs)
```

After looking at the plots, do you think that the top GO term from the table above is the most interesting? Which terms, if any, might be more informative of the biological mechanisms that drive motility differences between the two cell lines? Why?

```{r}
# fill in the name of your "favorite" GO term from the plot (use the values
# on the y-axis) — this could be the same or different from the 'go_description'
# you reported above
my_fav_go_term <- ""

# fill in your reason for choosing this GO term
my_rationale <- ""
```


## Submitting the prediction

You're now ready to submit the prediction. Just run the chunk below, a file with your prediction will be uploaded to Synapse and submitted to the challenge. You'll be able to see the results of your prediction on the mini-DREAM scoreboards, with the submission ID that gets printed below.

```{r}
library(synapser)

# If you didn't submit an answer for a previous module, remove the '#'s
# on the next 2 lines and enter your Synapse username and password
# before running this chunk.

# synLogin('my_synapse_username', 'my_synapse_password',
#          rememberMe = TRUE, silent = TRUE)

# If you submitted for Module 0, 1, 2, or 3, this should work.
synLogin(silent = TRUE)
submission <- submit_module_answers(module = 4)
```

Congrats — you’ve reached the end of **Module 4**! You can now return to the **mini-DREAM Challenge** site on Synapse.

---

# Bonus — Thought questions

**Even MORE fun:**  How many genes are in the sets we analyzed?  What would be the consequences of including more or less genes in our sets?  What are the potential perils of having gene sets that are "too small" or "too large"?  Would you expect to obtain many statistical enrichments for a small set?  What might result when the set is too large

